src/rust/src/lib.rs

```diff
@@ -1,132 +1,224 @@
 use extendr_api::prelude::*;
-use std::collections::HashMap;
+use std::iter::zip;
 
 #[extendr]
 fn model_matrix(data: List) -> Result<Robj> {
     let nrow = data.iter().next().map(|(_, col)| col.len()).unwrap_or(0);
-    let mut processed_columns: Vec<Array2<f64>> = Vec::new();
-    let mut column_names: Vec<String> = Vec::new();
+    let columns = data.iter().map(|(id, col)| {
+        if col.is_string() {
+            (id, R!("factor({{col}})").unwrap())
+        } else {
+            (id, col)
+        }
+    });
+
+    // every factor is turned into a one-hot vector, thus every factor results in levels() many more columns
+    let ncol: usize = columns
+        .clone()
+        .map(|(_id, x)| {
+            if x.is_factor() {
+                x.levels().unwrap().len() - 1
+            } else {
+                1
+            }
+        })
+        .sum();
+
+    // add intercept
+    let ncol = ncol + 1;
+
+    // note that we need one more column for the _intercept_
+    let mut processed_columns_matrix: RMatrix<f64> = RMatrix::new(nrow, ncol);
+    let processed_columns = processed_columns_matrix.as_real_slice_mut().unwrap();
+    let mut column_names: Vec<String> = Vec::with_capacity(ncol);
```

Comment:
Since we know how many extra columns, we can pre-allocate this. But that's not going to help much.. it does look nice.